cicecore/cicedynB/infrastructure/ice_grid.F90

```diff
@@ -1872,154 +2105,381 @@ subroutine Tlatlon
+         case('N2US')
+            call grid_average_X2YS('E' ,work1,narea,npm,work2tmp)
+         case('N2ES')
+            call grid_average_X2YS('SE',work1,narea,npm,work2tmp)
+
+         case default
+            call abort_ice(subname//'ERROR: unknown X2Y '//trim(X2Y))
+      end select
+
+      if (present(work2)) then
+         work2 = work2tmp
+      else
+         work1 = work2tmp
+      endif
 
-      end subroutine t2ugrid_vector
+      end subroutine grid_average_X2Y
 
 !=======================================================================
-
-! Shifts quantities from the T-cell midpoint (work1) to the U-cell
-! midpoint (work2)
+! Shifts quantities from one grid to another
+! State masked version, simple area weighted averager
 ! NOTE: Input array includes ghost cells that must be updated before
 !       calling this routine.
 !
-! author: Elizabeth C. Hunke, LANL
+! author: T. Craig
 
-      subroutine to_ugrid(work1,work2)
+      subroutine grid_average_X2YS(dir,work1,area1,mask1,work2)
 
-      use ice_constants, only: c0, p25
+      use ice_constants, only: c0
+
+      character(len=*) , intent(in) :: &
+         dir
 
       real (kind=dbl_kind), intent(in) :: &
-         work1(nx_block,ny_block,max_blocks)
+         work1(nx_block,ny_block,max_blocks), &
+         area1(nx_block,ny_block,max_blocks), &
+         mask1(nx_block,ny_block,max_blocks)
 
       real (kind=dbl_kind), intent(out) :: &
          work2(nx_block,ny_block,max_blocks)
 
-      type (block) :: &
-         this_block           ! block information for current block
-
-      character(len=*), parameter :: subname = '(to_ugrid)'
-
       ! local variables
 
       integer (kind=int_kind) :: &
          i, j, iblk, &
          ilo,ihi,jlo,jhi      ! beginning and end of physical domain
 
-      work2(:,:,:) = c0
-
-      !$OMP PARALLEL DO PRIVATE(iblk,i,j,ilo,ihi,jlo,jhi,this_block)
-      do iblk = 1, nblocks
-         this_block = get_block(blocks_ice(iblk),iblk)
-         ilo = this_block%ilo
-         ihi = this_block%ihi
-         jlo = this_block%jlo
-         jhi = this_block%jhi
-
-         do j = jlo, jhi
-         do i = ilo, ihi
-            work2(i,j,iblk) = p25 * &
-                              (work1(i,  j,  iblk)*tarea(i,  j,  iblk)  &
-                             + work1(i+1,j,  iblk)*tarea(i+1,j,  iblk)  &
-                             + work1(i,  j+1,iblk)*tarea(i,  j+1,iblk)  &
-                             + work1(i+1,j+1,iblk)*tarea(i+1,j+1,iblk)) &
-                             / uarea(i,  j,  iblk)
-         enddo
-         enddo
-      enddo
-      !$OMP END PARALLEL DO
-
-      end subroutine to_ugrid
-
-!=======================================================================
-
-! Transfer from U-cell centers to T-cell centers. Writes work into
-! another array that has ghost cells
-! NOTE: Input array is dimensioned only over physical cells.
-!
-! author: Elizabeth C. Hunke, LANL
-
-      subroutine u2tgrid_vector (work)
```

Comment:
The "_vector" suffix would indicate to me that this (old) routine would calculate vector averages, which it does not.  I'm guessing that it was called this because vector quantities are on the U grid (for the B-grid case), which needs the extra halo update before shifting.  (I looked at the 'blame' function and this one is on me...  sometime between 8 and 15 years ago).  This PR replaces this subroutine with "grid_average*", which is more appropriately named and should reproduce the current functionality on B grids.  As part of the C-grid development and testing, it would be good to check whether/how much doing proper vector averages would make a difference in the solution.  I suspect a lot, since the difference between B- and C-/CD-grids is all about the vector quantities.

Reply:
The only different between the to_u and t2ugrid_vector methods was a halo update.  So now the halo updated is required to be done outside the averager and the 4 old averages are all gone, but reimplemented in a bfb way in grid_average_X2Y (we can change the name if folks don't like that one).

Reply:
The grid_average_X2Y name is fine -- it's better than the previous subroutine name.  My main point here is that I don't think we're doing vector averages correctly in old or new code.